File: go/problem-3/problem-3.go
package main

import (
    "fmt"
    "math"
    "sort"
)

func main() {
    answer := greatestPrimeFactor(600851475143)
    fmt.Println(answer)
}

func greatestPrimeFactor(n int) int {
    pf := primeFactors(n)
    return pf[len(pf)-1]
}

func primeFactors(n int) []int {
    f := factors(n)
    var pf []int

    for _,factor := range f {
        if isPrime(factor) {
            pf = append(pf, factor)
        }
    }

    return pf
}

func factors(n int) []int {
    fsrN := flooredSquareRoot(n)
    var f []int

    for i := 1; i <= fsrN; i++ {
        if (n % i == 0) {
            if (n / i == i) {
                f = append(f, i)
            } else {
                f = append(f, i, n/i)
            }
        }
    }

    sort.Ints(f)
    return f
}

func isPrime(n int) bool {
    fsrN := flooredSquareRoot(n)

    for i := 2; i <= fsrN; i++ {
        if (n % i == 0) {
            return false
        }
    }

    return true
}

func flooredSquareRoot(n int) int {
    return int(math.Sqrt(float64(n)))
}

File: java/problem-3/Problem3.java
import java.lang.Math;
import java.util.*;

public class Problem3 {
    public static void main(String[] args) {
        long answer = greatestPrimeFactor(600851475143L);
        System.out.println(answer);
    }

    private static long greatestPrimeFactor(long n) {
        TreeSet<Long> primeFactors = primeFactors(n);
        return primeFactors.last();
    }

    private static TreeSet<Long> primeFactors(long n) {
        TreeSet<Long> primeFactors = new TreeSet<Long>();
        ArrayList<Long> factors = factors(n);

        for (Long factor : factors) {
            if(isPrime(factor)) {
                primeFactors.add(factor);
            }
        }

        return primeFactors;
    }

    private static ArrayList<Long> factors(long n) {
        ArrayList<Long> factors = new ArrayList<Long>();

        for(long i = 1; i <= Math.sqrt(n); i++) {
            if(n % i == 0) {
                long complement = n / i;

                factors.add(i);

                if(i != complement) {
                    factors.add(complement);
                }
            }
        }

        return factors;
    }

    private static boolean isPrime(long n) {
        for(long i = 2; i <= Math.sqrt(n); i++) {
            if(n % i == 0) {
                return false;
            }
        }
        return true;
    }
}
